### ISSUE TITLE
ResourceExhaustedError when formating ImageNet data to TFRecord

### ISSUE THREAD

**Issue by hfutxrg** [tensorflow/tensorflow#4685]
Hi All Tensorflow users,

I followed the steps in https://github.com/tensorflow/models/tree/master/inception to build the imagenet to TFRecord format. But when I execute the command 
"bazel-bin/inception/download_and_preprocess_imagenet "${DATA_DIR}"", 
I have the following errors:

Exception in thread Thread-6:
Traceback (most recent call last):
  File "/usr/lib64/python2.7/threading.py", line 811, in __bootstrap_inner
  File "/usr/lib64/python2.7/threading.py", line 764, in run
  File "/home/rengan/.cache/bazel/_bazel_rengan/2c9ddf610cf302ceda3f2b852669382c/execroot/inception/bazel-out/local-fastbuild/bin/inception/build_imagenet_data.runfiles/inception/inception/data/build_imagenet_data.py", line 388, in _process_image_files_batch
    image_buffer, height, width = _process_image(filename, coder)
  File "/home/rengan/.cache/bazel/_bazel_rengan/2c9ddf610cf302ceda3f2b852669382c/execroot/inception/bazel-out/local-fastbuild/bin/inception/build_imagenet_data.runfiles/inception/inception/data/build_imagenet_data.py", line 316, in _process_image
    image_data = tf.gfile.FastGFile(filename, 'r').read()
  File "/usr/lib/python2.7/site-packages/tensorflow/python/lib/io/file_io.py", line 102, in read
  File "/usr/lib/python2.7/site-packages/tensorflow/python/lib/io/file_io.py", line 72, in _preread_check
  File "/usr/lib64/python2.7/contextlib.py", line 24, in __exit__
  File "/usr/lib/python2.7/site-packages/tensorflow/python/framework/errors.py", line 463, in raise_exception_on_not_ok_status
ResourceExhaustedError: /cm/shared/scratch/rengan/Data/imagenet-data/raw-data/train//n03124170/n03124170_675.JPEG

There are similar errors in other threads. It successfully generated 128 validation data but less than 100 training data (should be 1024).

### Environment info
Operating System: Redhat 7.2

Installed version of CUDA and cuDNN: 
(please attach the output of `ls -l /path/to/cuda/lib/libcud*`):
-rw-r--r-- 1 root root   560184 May  8 02:00 /usr/local/cuda-8.0/lib64/libcudadevrt.a
lrwxrwxrwx 1 root root       16 Sep 12 15:13 /usr/local/cuda-8.0/lib64/libcudart.so -> libcudart.so.8.0
lrwxrwxrwx 1 root root       19 Sep 12 15:13 /usr/local/cuda-8.0/lib64/libcudart.so.8.0 -> libcudart.so.8.0.27
-rwxr-xr-x 1 root root   394472 May  8 02:00 /usr/local/cuda-8.0/lib64/libcudart.so.8.0.27
-rw-r--r-- 1 root root   737516 May  8 02:00 /usr/local/cuda-8.0/lib64/libcudart_static.a
-rwxr-xr-x 1 root root 60696704 Sep 24 09:12 /usr/local/cuda-8.0/lib64/libcudnn.so
-rwxr-xr-x 1 root root 60696704 Sep 24 09:12 /usr/local/cuda-8.0/lib64/libcudnn.so.5
-rwxr-xr-x 1 root root 60696704 Sep 24 09:12 /usr/local/cuda-8.0/lib64/libcudnn.so.5.1.3
-rw-r--r-- 1 root root 59715990 Sep 24 09:12 /usr/local/cuda-8.0/lib64/libcudnn_static.a

If installed from source, provide 
1. The commit hash (`git rev-parse HEAD`): f98c5ded31d7da0c2d127c28b2c16f0307a368f0
2. The output of `bazel version`: 
WARNING: Output base '/home/rengan/.cache/bazel/_bazel_rengan/2c9ddf610cf302ceda3f2b852669382c' is on NFS. This may lead to surprising failures and undetermined behavior.
.
Build label: 0.3.1- (@non-git)
Build target: bazel-out/local-fastbuild/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Thu Sep 29 22:19:27 2016 (1475187567)
Build timestamp: 1475187567
Build timestamp as int: 1475187567

Thanks in advance.

**Comment by huttunensami** [comment]
Hi,

I have a same kind of issue when trying to download and preprocess the MSCOCO dataset used in im2txt model  (https://github.com/tensorflow/models/tree/master/im2txt). It seems that the script `bazel-bin/im2txt/download_and_preprocess_mscoco "${MSCOCO_DIR}"` is able to process up to 1024 images but after that all the threads fail:
```
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/lib64/python2.7/threading.py", line 811, in __bootstrap_inner
  File "/usr/lib64/python2.7/threading.py", line 764, in run
  File "bazel-bin/im2txt/download_and_preprocess_mscoco.runfiles/im2txt/im2txt/data/build_mscoco_data.py", line 278, in _process_image_files
    sequence_example = _to_sequence_example(image, decoder, vocab)
  File "bazel-bin/im2txt/download_and_preprocess_mscoco.runfiles/im2txt/im2txt/data/build_mscoco_data.py", line 214, in _to_sequence_example
    encoded_image = f.read()
  File "<PYTHONDIR>/lib/python2.7/site-packages/tensorflow/python/lib/io/file_io.py", line 99, in read
  File "<PYTHONDIR>/lib/python2.7/site-packages/tensorflow/python/lib/io/file_io.py", line 72, in _preread_check
  File "/usr/lib64/python2.7/contextlib.py", line 24, in __exit__
  File "<PYTHONDIR>/lib/python2.7/site-packages/tensorflow/python/framework/errors.py", line 463, in raise_exception_on_not_ok_status
    pywrap_tensorflow.TF_GetCode(status))
ResourceExhaustedError: <DATA_DIR>/models/im2txt/im2txt/data/mscoco/raw-data/train2014/COCO_train2014_000000069569.jpg
```
Apparently the script does not close the file handles it uses properly which leads to this kind of unfortunate situation.

Operating System: CentOS Linux release 7.2.1511 (Core)
CUDA 7.5
cuDNN 5.0
TensorFlow master branch

**Comment by cshallue** [comment]
Ah, looks like there is a missing call to writer.close() in models/im2txt/im2txt/data/build_mscoco_data.py
I'll send out a fix shortly.

**Comment by hfutxrg** [comment]
Thanks for pointing out this issue. I added the writer.close() in models/inception/inception/data/build_imagenet_data.py and the problem was solved.

**Comment by siavashk** [comment]
@cshallue can you tell us how to patch things while you work on a fix?

**Comment by cshallue** [comment]
I have pushed the fix to this branch:
https://github.com/cshallue/models/tree/writer

I will merge into tensorflow_models/master soon.

**Comment by siavashk** [comment]
@cshallue if all that is different is adding writer.close on line 289 of build_mscoco_data.py, then this is not solving it for me. I am having the same exact problem as @huttunensami.

**Comment by cshallue** [comment]
Have you tried reducing the number of threads?

**Comment by siavashk** [comment]
@cshallue I am running with 1 thread by changing line 133 in build_mscoco_data.py

tf.flags.DEFINE_integer("num_threads", 1,
                        "Number of threads to preprocess the images.")

**Comment by cshallue** [comment]
I've pushed the patch to call writer.close(), which fixes the issue for @hfutxrg

I'm not sure yet why the issue still exists for @siavashk

**Comment by siavashk** [comment]
@cshallue I will continue digging for the next hour to see if I can get it to work, and if I do, I will reply on the same thread. In case it helps, I am building this on a cloud instance with a K2. The compute capability of these cards is 3.0, which is what I configured Tensorflow with.

**Comment by cshallue** [comment]
Looking more closely at the stack trace above, it's clear that the error is thrown while reading an image from disk:
  with tf.gfile.FastGFile(image.filename, "r") as f:
    encoded_image = f.read()

It sounds like you've run out of memory. Is there any way you can increase that on your cloud instance?

**Comment by siavashk** [comment]
I have a GPU instance from Softlayer, the available memory is 64GBs. Even though it is highly unlikely, I checked this before posting on the thread using top.

**Comment by cshallue** [comment]
I wonder if you get still get an error if you use the native Python open() method instead of tf.gfile.FastGFile()?

**Comment by siavashk** [comment]
Good point. I tried it, processing is slow but seems to be working so far with 8 thread. Previously it would fail immediately. Did we just find a bug in Tensorflow IO?

I tensorflow/core/common_runtime/gpu/gpu_device.cc:972] DMA: 0 1 
I tensorflow/core/common_runtime/gpu/gpu_device.cc:982] 0:   Y Y 
I tensorflow/core/common_runtime/gpu/gpu_device.cc:982] 1:   Y Y 
I tensorflow/core/common_runtime/gpu/gpu_device.cc:1041] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GRID K2, pci bus id: 0000:83:00.0)
I tensorflow/core/common_runtime/gpu/gpu_device.cc:1041] Creating TensorFlow device (/gpu:1) -> (device: 1, name: GRID K2, pci bus id: 0000:84:00.0)
Launching 8 threads for spacings: [[0, 73296], [73296, 146592], [146592, 219888], [219888, 293184], [293184, 366480], [366480, 439776], [439776, 513072], [513072, 586368]]
2016-09-30 17:54:34.244437 [thread 0]: Processed 1000 of 73296 items in thread batch.
2016-09-30 17:54:56.926579 [thread 4]: Processed 1000 of 73296 items in thread batch.
2016-09-30 17:55:03.010200 [thread 1]: Processed 1000 of 73296 items in thread batch.
2016-09-30 17:55:03.078273 [thread 6]: Processed 1000 of 73296 items in thread batch.
2016-09-30 17:55:06.125806 [thread 5]: Processed 1000 of 73296 items in thread batch.
2016-09-30 17:55:06.261466 [thread 3]: Processed 1000 of 73296 items in thread batch.
2016-09-30 17:55:06.557369 [thread 2]: Processed 1000 of 73296 items in thread batch.
2016-09-30 17:55:06.630173 [thread 7]: Processed 1000 of 73296 items in thread batch.

**Comment by cshallue** [comment]
This might indeed be a bug in tf.gfile.FastGFile().
Please let me know if it finishes successfully for you now (the preprocessing usually takes 1+ hours).

**Comment by siavashk** [comment]
@cshallue It worked!

**Comment by cshallue** [comment]
Great! I will push a patch shortly to unblock everyone, and I'll close this issue.
I'll follow up next week about tf.gfile.FastGFile(), and work on that in a separate issue.
